README.md
# Hamsket <!-- omit in toc -->

<p align="center">
<a href="https://github.com/TheGoddessInari/hamsket"><img src="./resources/Icon.png" width="256px" alt="Hamsket"></a>
</p>

#### Free, Open Source and Cross Platform messaging and emailing app that combines common web applications into one. <!-- omit in toc -->

[![GNU GPL v3][license]](https://www.gnu.org/licenses/gpl-3.0.en.html)
[![Gitter][gitter]](https://gitter.im/TheGoddessInari/hamsket)
[![Release][release]](https://github.com/TheGoddessInari/hamsket/releases/latest)
[![Nightly][nightly]](https://github.com/TheGoddessInari/hamsket/releases/tag/nightly)

[![DevOps Build][devops-build]](https://dev.azure.com/TheGoddessInari/hamsket/_build?definitionId=2)
[![DevOps Deployment][devops-deploy]](https://dev.azure.com/TheGoddessInari/hamsket/_release?definitionId=1)
[![Dependencies][dependencies]](https://david-dm.org/TheGoddessInari/hamsket)
[![DevDependencies][dev-dependencies]](https://david-dm.org/TheGoddessInari/hamsket?type=dev)

[license]: https://img.shields.io/github/license/TheGoddessInari/hamsket.svg
[gitter]: https://img.shields.io/gitter/room/TheGoddessInari/hamsket.svg
[release]: https://img.shields.io/github/release/TheGoddessInari/hamsket.svg
[nightly]: https://img.shields.io/badge/Nightly-tag-red

[devops-build]: https://img.shields.io/azure-devops/build/TheGoddessInari/f5a88978-10c7-4069-97e0-59b9fc6d36a0/2.svg?logo=azure-devops
[devops-deploy]: https://img.shields.io/azure-devops/release/TheGoddessInari/f5a88978-10c7-4069-97e0-59b9fc6d36a0/1/1.svg?logo=azure-devops
[dependencies]: https://img.shields.io/david/TheGoddessInari/hamsket.svg
[dev-dependencies]: https://img.shields.io/david/dev/TheGoddessInari/hamsket.svg

##### Nightly build artifacts are available in Windows ([Win64](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket.Setup.0.6.5.exe)), Mac ([DMG](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket-0.6.5.dmg) / [App](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket-0.6.5-mac.zip)), and Linux ([AppImage64](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket-0.6.5.AppImage) / [RPM64](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket-0.6.5.x86_64.rpm) / [DEB64](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket_0.6.5_amd64.deb) / [Tarball64](https://github.com/TheGoddessInari/hamsket/releases/download/nightly/Hamsket-0.6.5.tar.gz)) packaging flavors. <!-- omit in toc -->

###### Logo designed by [Andriy Yurchenko](http://andyur.com/). <!-- omit in toc -->

###### Original Rambox by [Ramiro Saenz](https://github.com/saenzramiro). <!-- omit in toc -->

----------

## Table of Contents

- [Table of Contents](#table-of-contents)
- [Screenshot](#screenshot)
- [Services available - 95](#services-available---95)
- [Features](#features)
- [Privacy](#privacy)
- [Donations](#donations)
- [Compiling](#compiling)
- [Contributing](#contributing)
- [Quickstart](#quickstart)
- [Disclosure](#disclosure)
- [License](#license)

----------

## Screenshot

![Hamsket](./resources/screenshots/win1.png)

----------

## Services available - 95

<img width="80" align="left" src="./resources/icons/whatsapp.png" alt="WhatsApp" title="WhatsApp">
<img width="80" align="left" src="./resources/icons/messenger.png" alt="Messenger" title="Messenger">
<img width="80" align="left" src="./resources/icons/skype.png" alt="Skype" title="Skype">
<img width="80" align="left" src="./resources/icons/slack.png" alt="Slack" title="Slack">
<img width="80" align="left" src="./resources/icons/hangouts.png" alt="Hangouts" title="Hangouts">
<img width="80" align="left" src="./resources/icons/telegram.png" alt="Telegram" title="Telegram">
<img width="80" align="left" src="./resources/icons/wechat.png" alt="WeChat" title="WeChat">
<img width="80" align="left" src="./resources/icons/gmail.png" alt="Gmail" title="Gmail">
<img width="80" align="left" src="./resources/icons/hipchat.png" alt="HipChat" title="HipChat">

<img width="80" align="left" src="./resources/icons/chatwork.png" alt="ChatWork" title="ChatWork">
<img width="80" align="left" src="./resources/icons/groupme.png" alt="GroupMe" title="GroupMe">
<img width="80" align="left" src="./resources/icons/grape.png" alt="Grape" title="Grape">
<img width="80" align="left" src="./resources/icons/gitter.png" alt="Gitter" title="Gitter">
<img width="80" align="left" src="./resources/icons/steam.png" alt="Steam" title="Steam">
<img width="80" align="left" src="./resources/icons/discord.png" alt="Discord" title="Discord">
<img width="80" align="left" src="./resources/icons/noysi.png" alt="Noysi" title="Noysi">
<img width="80" align="left" src="./resources/icons/outlook.png" alt="Outlook" title="Outlook">
<img width="80" align="left" src="./resources/icons/outlook365.png" alt="Outlook 365" title="Outlook 365">
<img width="80" align="left" src="./resources/icons/tutanota.png" alt="TutaNota" title="TutaNota">

<img width="80" align="left" src="./resources/icons/hushmail.png" alt="Hushmail" title="Hushmail">
<img width="80" align="left" src="./resources/icons/bearychat.png" alt="BearyChat" title="BearyChat">
<img width="80" align="left" src="./resources/icons/aol.png" alt="Aol" title="Aol">
<img width="80" align="left" src="./resources/icons/sync.png" alt="Sync" title="Sync">
<img width="80" align="left" src="./resources/icons/wire.png" alt="Wire" title="Wire">
<img width="80" align="left" src="./resources/icons/rocketchat.png" alt="Rocket Chat" title="Rocket Chat">
<img width="80" align="left" src="./resources/icons/missive.png" alt="Missive" title="Missive">
<img width="80" align="left" src="./resources/icons/yahoo.png" alt="Yahoo! Mail" title="Yahoo! Mail">
<img width="80" align="left" src="./resources/icons/ryver.png" alt="Ryver" title="Ryver">

<img width="80" align="left" src="./resources/icons/yandex.png" alt="Yandex Mail" title="Yandex Mail">
<img width="80" align="left" src="./resources/icons/dasher.png" alt="Dasher" title="Dasher">
<img width="80" align="left" src="./resources/icons/dingtalk.png" alt="DingTalk" title="DingTalk">
<img width="80" align="left" src="./resources/icons/flowdock.png" alt="FlowDock" title="FlowDock">
<img width="80" align="left" src="./resources/icons/mattermost.png" alt="Mattermost" title="Mattermost">
<img width="80" align="left" src="./resources/icons/voxer.png" alt="Voxer" title="Voxer">
<img width="80" align="left" src="./resources/icons/glip.png" alt="Glip" title="Glip">
<img width="80" align="left" src="./resources/icons/mysms.png" alt="mysms" title="mysms">
<img width="80" align="left" src="./resources/icons/icq.png" alt="ICQ" title="ICQ">

<img width="80" align="left" src="./resources/icons/tweetdeck.png" alt="TweetDeck" title="TweetDeck">
<img width="80" align="left" src="./resources/icons/zinc.png" alt="Zinc" title="Zinc">
<img width="80" align="left" src="./resources/icons/freenode.png" alt="FreeNode" title="FreeNode">
<img width="80" align="left" src="./resources/icons/mightytext.png" alt="MightyText" title="MightyText">
<img width="80" align="left" src="./resources/icons/zohoemail.png" alt="Zoho Email" title="Zoho Email">
<img width="80" align="left" src="./resources/icons/zohochat.png" alt="Zoho Chat" title="Zoho Chat">
<img width="80" align="left" src="./resources/icons/roundcube.png" alt="Roundcube" title="Roundcube">
<img width="80" align="left" src="./resources/icons/horde.png" alt="Horde" title="Horde">
<img width="80" align="left" src="./resources/icons/squirrelmail.png" alt="SquirrelMail" title="SquirrelMail">
<img width="80" align="left" src="./resources/icons/zimbra.png" alt="Zimbra" title="Zimbra">

<img width="80" align="left" src="./resources/icons/hootsuite.png" alt="Hootsuite" title="Hootsuite">
<img width="80" align="left" src="./resources/icons/amium.png" alt="Amium" title="Amium">
<img width="80" align="left" src="./resources/icons/rainloop.png" alt="RainLoop" title="RainLoop">
<img width="80" align="left" src="./resources/icons/icloud.png" alt="iCloud Mail" title="iCloud Mail">
<img width="80" align="left" src="./resources/icons/irccloud.png" alt="IRC Cloud" title="IRC Cloud">
<img width="80" align="left" src="./resources/icons/kiwi.png" alt="Kiwi IRC" title="Kiwi IRC">
<img width="80" align="left" src="./resources/icons/glowingbear.png" alt="Glowing Bear" title="Glowing Bear">
<img width="80" align="left" src="./resources/icons/smooch.png" alt="Smooch" title="Smooch">
<img width="80" align="left" src="./resources/icons/crisp.png" alt="Crisp" title="Crisp">
<img width="80" align="left" src="./resources/icons/flock.png" alt="Flock" title="Flock">
<img width="80" align="left" src="./resources/icons/openmailbox.png" alt="Openmailbox" title="Openmailbox">

<img width="80" align="left" src="./resources/icons/typetalk.png" alt="Typetalk" title="Typetalk">
<img width="80" align="left" src="./resources/icons/drift.png" alt="Drift" title="Drift">
<img width="80" align="left" src="./resources/icons/webexteams.png" alt="Cisco Webex Teams" title="Cisco Webex Teams">
<img width="80" align="left" src="./resources/icons/fleep.png" alt="Fleep" title="Fleep">
<img width="80" align="left" src="./resources/icons/socialcast.png" alt="Socialcast" title="Socialcast">
<img width="80" align="left" src="./resources/icons/element.png" alt="Element" title="Element">
<img width="80" align="left" src="./resources/icons/pushbullet.png" alt="Pushbullet" title="Pushbullet">
<img width="80" align="left" src="./resources/icons/movim.png" alt="Movim" title="Movim">

<img width="80" align="left" src="./resources/icons/kaiwa.png" alt="Kaiwa" title="Kaiwa">
<img width="80" align="left" src="./resources/icons/zyptonite.png" alt="Zyptonite" title="Zyptonite">
<img width="80" align="left" src="./resources/icons/linkedin.png" alt="LinkedIn" title="LinkedIn">
<img width="80" align="left" src="./resources/icons/lounge.png" alt="The Lounge" title="The Lounge">
<img width="80" align="left" src="./resources/icons/kezmo.png" alt="Kezmo" title="Kezmo">
<img width="80" align="left" src="./resources/icons/teams.png" alt="Teams" title="Teams">
<img width="80" align="left" src="./resources/icons/xing.png" alt="Xing" title="Xing">
<img width="80" align="left" src="./resources/icons/workplace.png" alt="Workplace" title="Workplace">
<img width="80" align="left" src="./resources/icons/fastmail.png" alt="FastMail" title="FastMail">
<img width="80" align="left" src="./resources/icons/hibox.png" alt="Hibox" title="Hibox">
<img width="80" align="left" src="./resources/icons/jandi.png" alt="Jandi" title="Jandi">
<img width="80" align="left" src="./resources/icons/threema.png" alt="Threema" title="Threema">
<img width="80" align="left" src="./resources/icons/messengerpages.png" alt="Messenger for Pages" title="Messenger for Pages">
<img width="80" align="left" src="./resources/icons/vk.png" alt="VK Messenger" title="VK Messenger">
<img width="80" align="left" src="./resources/icons/mastodon.png" alt="Mastodon" title="Mastodon">
<img width="80" align="left" src="./resources/icons/teamworkchat.png" alt="Teamwork Chat" title="Teamwork Chat">
<img width="80" align="left" src="./resources/icons/clocktweets.png" alt="ClockTweets" title="ClockTweets">
<img width="80" align="left" src="./resources/icons/intercom.png" alt="Intercom" title="Intercom">
<img width="80" align="left" src="./resources/icons/googlevoice.png" alt="Voice" title="Voice">
<img width="80" align="left" src="./resources/icons/sandstorm.png" alt="Sandstorm" title="Sandstorm">
<img width="80" align="left" src="./resources/icons/gadugadu.png" alt="Gadu-Gadu" title="Gadu-Gadu">
<img width="80" align="left" src="./resources/icons/mailru.png" alt="Mail.Ru" title="Mail.Ru">
<img width="80" align="left" src="./resources/icons/kune.png" alt="Kune" title="Kune">
<img width="80" align="left" src="./resources/icons/zulip.png" alt="Zulip" title="Zulip">
<img width="80" align="left" src="./resources/icons/stride.png" alt="Stride" title="Stride">
<img width="80" align="left" src="./resources/icons/hangoutschat.png" alt="Hangouts Chat" title="Hangouts Chat">
<img width="80" align="left" src="./resources/icons/messengerpages.png" alt="Messenger for Business" title="Messenger for Business">
<img width="80" align="left" src="./resources/icons/androidmessages.png" alt="Android Messages" title="Android Messages">
<img width="80" align="left" src="./resources/icons/instagram.png" alt="Instagram" title="Instagram">

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

----------

## Features

- [ ] Multi-language.
- [ ] Sync your configuration between multiple computers.
- [x] Master Password.
- [x] Lock Hamsket if you will be away for a period of time.
- [x] Don't disturb mode.
- [x] Reorder applications in the tab bar.
- [x] Notification badge in the tab.
- [x] Minimize to tray.
- [x] Mute audio to specific service.
- [x] Separate tabs floating to the right.
- [x] Disable a service instead of remove it.
- [x] Start automatically on system startup.
- [x] Custom Javascript injection.
- [x] Custom CSS injection.
- [x] Keyboard Shortcuts.
- [x] Proxy.
- [x] Passive mouse/touch event listeners by default.
- [x] Lowered timer granularity to save CPU time by default.
- [x] No user tracking.
- [x] Fully open source, won't change
- [x] Highly responsive to issues, PRs, and help.
- [x] Offline backup and restore.

----------

## Privacy

No personal information will be saved or tracked.

Sessions will persist using the [partition:persist](https://electronjs.org/docs/api/webview-tag#partition) attribute for Webviews. So every time you open Hamsket, your sessions will keep alive until you remove the service.

~~Sync feature use [Auth0](https://auth0.com/) for Single Sign On & Token Based Authentication and to store the services that user is using (and the configuration for each service). You are always welcome to check the code! ;)~~

Auth0 is disabled pending name change and rewrite of the functionality to best preserve your security and functionality.

----------

## Donations

Keep in mind, I'm not pushing this. I'll do my best regardless.

| Type                |                    URL/Wallet                    |
| --------------------| :----------------------------------------------: |
| GitHub Sponsors     |       <https://github.com/sponsors/TheGoddessInari/>       |

----------

## Compiling

To build your own copy, you need:

- An 8.0 version of the [Java SE Development Kit](http://www.oracle.com/technetwork/java/javase/downloads/index.html).
- [Ruby](https://www.ruby-lang.org/en/downloads/) >2.3 && <3.0.
- [Sencha CMD](https://www.sencha.com/products/extjs/cmd-download/) 6.7.0 or newer, and to install it with the Compass extension.
- [NodeJS](https://nodejs.org/en/download/) 14.x or newer with npm.

These **need** to be added to your PATH.

```shell
git clone https://github.com/TheGoddessInari/hamsket.git
cd hamsket
npm install
npm run repack:win64
```

You can substitute win64 for linux32, linux64, win32, or osx as desired. This will create an unpacked directory in '**dist**' for testing.

```shell
npm run setup
```

This step instead of **repack** will create an installer appropriate to your platform in '**dist**'.
You can place unpacked files over where the installers normally place their files for development/testing purposes.

**NOTE:** 32-bit builds are **deprecated** by electron-builder and are likely to be removed at some point in the future.

----------

## [Contributing](./CONTRIBUTING.md)

Want to report a bug, request a feature, contribute to or translate Hamsket? We need all the help we can get! Fork and work!

----------

## Quickstart

```shell
git clone https://github.com/TheGoddessInari/hamsket.git
cd hamsket
npm install
npm start
```

See [Contributing.md](./CONTRIBUTING.md) for more detailed information about getting set up.

----------

## Disclosure

Hamsket is not affiliated with any of the messaging apps offered.

----------

## License

[GNU GPL v3](./LICENSE)

----------


CHANGELOG.md
*In compliance with the [GPL-3.0](https://opensource.org/licenses/GPL-3.0) license: I declare that this version of the program contains my modifications, which can be seen through the usual "git" mechanism.*  


2022-08  
Contributor(s):  
TheGoddessInari  
>Bump version to 0.6.5.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2022-06  
Contributor(s):  
TheGoddessInari  
>Add Instagram.Fixes #330.  
>Update README.md for new service count.  
>Remove the obsolete GoFundMe links.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2022-03  
Contributor(s):  
TheGoddessInari  
>Bump version to 0.6.4.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2021-12  
Contributor(s):  
TheGoddessInari  
>Update nodejs requirement to >= 14, update electron to 13.6.3.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-12  
Contributor(s):  
TheGoddessInari  
>Bump version to 0.6.2.  
>Bump version to 0.6.3.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-11  
Contributor(s):  
TheGoddessInari  
>Bump version to 0.6.1.  
>Update to extjs 7. (broken)  
>Add a link to the nightly tag.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-07  
Contributor(s):  
TheGoddessInari  
>Matrix is rebranding Riot to Element and plans to shut down Riot "within a few weeks".Fixes #216.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-04  
Contributor(s):  
TheGoddessInari  
>Google Inbox was shut down on April 9th, 2019.  
>Fix README.md to use correct david-dm URL after package.json unsplit.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-02  
Contributor(s):  
TheGoddessInari  
>README: Fix dependencies link.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-01  
Contributor(s):  
TheGoddessInari  
Daniel Peukert  
>Merge pull request #158 from dpeukert/replace-deprecated-functionReplaced deprecated functions  
>Merge pull request #163 from tumido/add-android-messagesAdd Android Messages  
>Replaced deprecated .getName() with .name  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-09  
Contributor(s):  
TheGoddessInari  
>Remove Google Allo.Google Allo was retired on March 12, 2019 by Google.  
>Remove coin addresses.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-08  
Contributor(s):  
TheGoddessInari  
>Reimplement Ctrl/Cmd+1..9 shortcuts.Fixes #107.  
>Bump version in README and package-lock to 0.6.0.  
>Fix README. ;)  
>Fix AppImage link in README.md.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-07  
Contributor(s):  
TheGoddessInari  
>Rename baseline from Rambox-OS to Hamsket.  
>Rename repo to hamsket.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-06  
Contributor(s):  
TheGoddessInari  
>Reformat README.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-05  
Contributor(s):  
TheGoddessInari  
>Replace var with let or const as appropriate.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-04  
Contributor(s):  
TheGoddessInari  
>Update loadscreen to modernizr 3.6.0.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-03  
Contributor(s):  
TheGoddessInari  
>WebView: Several different cleanups. Fix zoom level logic. Restore zoom level on start.Now using getWebView consistently.Added getWebContents and setZoomLevel methods.Added blur and focus methods.  
>README.md: Update DevOps status to shields.io.  
>README: Consistently reformat.  
>README: Add deployment status badge.  
>Fix tab shortcuts for MacOS, add another shortcut to tab with.  
>Azure DevOps is building Windows, so change URL and remove appveyor.yml.  
>Add Cmd+Q shortcut as mentioned in #65.  
>Update Sencha bits to need 6.7.0, update README.Sencha CMD 6.7.0 finally fixes the problem with arrow functions,so they'll finally be able to be used in the source if desirable.  
>README.md: Redirect non-Windows links to 'nightly', add MacOS links.Fixes #65  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-02  
Contributor(s):  
TheGoddessInari  
>Add alt+left and alt+right shortcuts for switching tabs.  
>Use path.join for icons.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-01  
Contributor(s):  
Jakob Struye  
Álvaro Brey  
TheGoddessInari  
>Redo shortcuts via hidden menus, adds quit keybind, move F1 F2 keybinds.This fixes #66.Also adds Ctrl-Shift-Q to be able to quit via keyboard easily, butprevent accidents.Also move F1 and F2 keybinds to Alt+F1 and Alt+F2respectively to also prevent accidents.  
>Fix locale path to be app relative, like was done for icons.  
>README: fix feature list (#81)* README: fix feature list* Fix non-rendering checkbox  
>README: Note that Auth0 is disabled until after the rename.  
>Rebranded Cisco Spark to Cisco Webex Teams  
>Partially revert da64582cf3c0026a1b06b3a936f813d9b629a220. App.quit should work again.  
>Merge pull request #80 from JakobStruye/os-spark-to-webexRebranded Cisco Spark to Cisco Webex Teams  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2018-12  
Contributor(s):  
TheGoddessInari  
>Remove references to `app watch`, as this never made sense.Also remove mention of 'contributed services'.  
>Add git hashes to buildversion.Also replace 'rimraf' with the more generic 'shx'.  
>README: Somehow the AppImage link changed, fix it.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2018-11  
Contributor(s):  
TheGoddessInari  
>CONTRIBUTING.md: Let's clean this up.  
>Update DEB64 link.Closes #63  
>README.md: Remove reference to env.js.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2018-08  
Contributor(s):  
TheGoddessInari  

app/view/main/Main.js
Ext.define('Hamsket.view.main.Main', {
	 extend: 'Ext.tab.Panel'
	,requires: [
		 'Hamsket.view.main.MainController'
		,'Hamsket.view.main.MainModel'
		,'Hamsket.ux.WebView'
		,'Hamsket.ux.mixin.Badge'
		,'Hamsket.view.add.Add'
		,'Ext.ux.TabReorderer'
	]

	,xtype: 'app-main'

	,controller: 'main'
	,viewModel: {
		type: 'main'
	}

	,plugins: [
		{
			 ptype: 'tabreorderer'
		}
	]

	,autoRender: true
	,autoShow: true
	,deferredRender: false
	,tabBar: {
		items: [
			{
				 xtype: 'button'
				,html: '<span class="fa fa-user-plus" style="color:red;font-size:16px;cursor:pointer;padding:0 5px;"></span>'
				,baseCls: ''
				,tooltip: 'Sponsor'
				,href: 'https://github.com/sponsors/TheGoddessInari/'
			}
		]
	}
	,items: [
		{
			 icon: 'resources/IconTray@2x.png'
			,id: 'hamsketTab'
			,closable: false
			,reorderable: false
			,autoScroll: true
			,layout: 'hbox'
			,tabConfig: {} // Created empty for Keyboard Shortcuts
			,items: [
				{
					 xtype: 'panel'
					,title: locale['app.main[0]']
					,margin: '0 5 0 0'
					,flex: 2
					,header: { height: 50 }
					,tools: [
						{
							 xtype: 'checkboxgroup'
							,items: [
								{
									 xtype: 'checkbox'
									,boxLabel: locale['app.main[1]']
									,name: 'messaging'
									,checked: true
									,uncheckedValue: false
									,inputValue: true
								}
								,{
									 xtype: 'checkbox'
									,boxLabel: locale['app.main[2]']
									,margin: '0 10 0 10'
									,name: 'email'
									,checked: true
									,uncheckedValue: false
									,inputValue: true
								}
							]
							,listeners: {
								change: 'doTypeFilter'
							}
						}
						,{
							 xtype: 'textfield'
							,grow: true
							,growMin: 120
							,growMax: 170
							,triggers: {
								 clear: {
									 weight: 0
									,cls: Ext.baseCSSPrefix + 'form-clear-trigger'
									,hidden: true
									,handler: 'onClearClick'
								}
								,search: {
									 weight: 1
									,cls: Ext.baseCSSPrefix + 'form-search-trigger search-trigger'
								}
							}
							,listeners: {
								 change: 'onSearchServiceChange'
								,afterrender: 'onSearchRender'
								,specialkey: 'onSearchEnter'
							}
						}
					]
					,items: [
						{
							 xtype: 'dataview'
							,store: 'ServicesList'
							,itemSelector: 'div.service'
							,tpl: [
								 '<tpl for=".">'
									,'<div class="service" data-qtip="{description}">'
										,'<img src="resources/icons/{logo}" width="48" />'
										,'<span>{name}</span>'
									,'</div>'
								,'</tpl>'
							]
							,emptyText: '<div style="padding: 20px;">'+locale['app.main[3]']+'</div>'
							,listeners: {
								itemclick: 'onNewServiceSelect'
							}
						}
					]
				}
				,{
					 xtype: 'grid'
					,title: locale['app.main[4]']
					,store: 'Services'
					,hideHeaders: true
					,margin: '0 0 0 5'
					,flex: 1
					,header: { height: 50 }
					,features: [
						{
							 ftype:'grouping'
							,collapsible: false
							,groupHeaderTpl: '{columnName:uppercase}: {name:capitalize} ({rows.length} {[values.rows.length > 1 ? "'+locale['app.main[9]']+'" : "'+locale['app.main[8]']+'"]})'
						}
					]
					,plugins: {
						 ptype: 'cellediting'
						,clicksToEdit: 2
					}
					,tools: [
						{
							 xtype: 'button'
							,glyph: 'XF1F8@FontAwesome'
							,baseCls: ''
							,tooltip: locale['app.main[10]']
							,handler: 'removeAllServices'
						}
					]
					,columns: [
						{
							 xtype: 'templatecolumn'
							,width: 50
							,variableRowHeight: true
							,tpl: `<img src='{[ values.type !== "custom" ? "resources/icons/"+values.logo : (values.logo == "" ? "resources/icons/custom.png" : values.logo) ]}' data-qtip='{type:capitalize}' width='32' style='{[ values.enabled ? "-webkit-filter: grayscale(0)" : "-webkit-filter: grayscale(1)" ]}' />`
						}
						,{
							 dataIndex: 'name'
							,variableRowHeight: true
							,flex: 1
							,editor: {
								 xtype: 'textfield'
								,allowBlank: true
							},
							renderer: Ext.String.htmlEncode
						}
						,{
							 xtype: 'actioncolumn'
							,width: 60
							,align: 'right'
							,items: [
								{
									 glyph: 0xF1F7
									,tooltip: locale['app.main[11]']
									,getClass( value, metaData, record, rowIndex, colIndex, store, view ){
										if ( record.get('notifications') ) return 'x-hidden';
									}
								}
								,{
									 glyph: 0xF026
									,tooltip: locale['app.main[12]']
									,getClass( value, metaData, record, rowIndex, colIndex, store, view ){
										if ( !record.get('muted') ) return 'x-hidden';
									}
								}
							]
						}
						,{
							 xtype: 'actioncolumn'
							,width: 60
							,align: 'center'
							,items: [
								{
									 glyph: 0xF013
									,tooltip: locale['app.main[13]']
									,handler: 'configureService'
									,getClass(){ return 'x-hidden-display'; }
								}
								,{
									 glyph: 0xF1F8
									,tooltip: locale['app.main[14]']
									,handler: 'removeService'
									,getClass(){ return 'x-hidden-display'; }
								}
							]
						}
						,{
							 xtype: 'checkcolumn'
							,width: 40
							,dataIndex: 'enabled'
							,renderer(value, metaData) {
								metaData.tdAttr = 'data-qtip="Service '+(value ? 'Enabled' : 'Disabled')+'"';
								return this.defaultRenderer(value, metaData);
							}
							,listeners: {
								checkchange: 'onEnableDisableService'
							}
						}
					]
					,viewConfig: {
						 emptyText: locale['app.main[15]']
						,forceFit: true
						,stripeRows: true
					}
					,listeners: {
						 edit: 'onRenameService'
						,rowdblclick: 'showServiceTab'
					}
				}
			]
			,tbar: {
				 xtype: 'toolbar'
				,height: 42
				,ui: 'main'
				,enableOverflow: true
				,overflowHandler: 'menu'
				,items: [
					{
						 glyph: 'XF1F7@FontAwesome'
						,text: locale['app.main[16]']+': '+(JSON.parse(localStorage.getItem('dontDisturb')) ? locale['app.window[20]'] : locale['app.window[21]'])
						,tooltip: locale['app.main[17]']+'<br/><b>'+locale['app.main[18]']+': Alt+F1</b>'
						,enableToggle: true
						,handler: 'dontDisturb'
						,reference: 'disturbBtn'
						,id: 'disturbBtn'
						,pressed: JSON.parse(localStorage.getItem('dontDisturb'))
					}
					,{
						 glyph: 'XF023@FontAwesome'
						,text: locale['app.main[19]']
						,tooltip: locale['app.main[20]']+'<br/><b>'+locale['app.main[18]']+': Alt+F2</b>'
						,handler: 'lockHamsket'
						,id: 'lockHamsketBtn'
					}
					,'->'
					,{
						 text: 'Backup'
						,glyph: 'XF0C7@FontAwesome'
						,id: 'backupBtn'
						,tooltip: 'Backup services to disk'
						,scope: Hamsket.ux.FileBackup
						,handler: Hamsket.ux.FileBackup.backupConfiguration
					}
					,{
						text: 'Restore'
					   ,glyph: 'XF115@FontAwesome'
					   ,id: 'restoreBtn'
					   ,tooltip: 'Restore services from disk'
					   ,scope: Hamsket.ux.FileBackup
					   ,handler: Hamsket.ux.FileBackup.restoreConfiguration
				   }
				   ,{
						 tooltip: locale['preferences[0]']
						,glyph: 'XF013@FontAwesome'
						,handler: 'openPreferences'
					}
				]
			}
			,bbar: [
				{
					 xtype: 'segmentedbutton'
					,allowToggle: false
					,items: [
						{
							 text: '<b>Help out</b> with'
							,pressed: true
						}
						,{
							 text: 'Sponsor'
							,glyph: 'XF234@FontAwesome'
							,href: 'https://github.com/sponsors/TheGoddessInari'
						}
					]
					 /* TODO: Translation
						,{
							 text: 'Translation'
							,glyph: 'XF0AC@FontAwesome'
							,href: 'https://crowdin.com/project/hamsket/invite'
						}
					]*/
				}
				,'->'
				,{
					 xtype: 'label'
					,html: '<span class="fa fa-code-fork" style="color:black;"></span> '+locale['app.main[26]']+' <span class="fa fa-meh-o" style="color:red;"></span> '+locale['app.main[27]'].replace('USA', '<img src="resources/flag.png" alt="United States" data-qtip="United States" />')
				}
				,'->'
				,{
					xtype: 'segmentedbutton'
					,allowToggle: false
					,items: [
						{
							 text: '<b>Follow us</b>'
							,pressed: true
						}
						,{
							 glyph: 'XF082@FontAwesome'
							,href: 'https://www.facebook.com/TheGoddessInari'
						}
						,{
							 glyph: 'XF099@FontAwesome'
							,href: 'https://www.twitter.com/TheGoddessInari'
						}
						,{
							 glyph: 'XF09B@FontAwesome'
							,href: 'https://www.github.com/TheGoddessInari/hamsket'
						}
					]
				}
			]
		}
		,{ id: 'tbfill', tabConfig : { xtype : 'tbfill' } }
	]

	,listeners: {
		 tabchange: 'onTabChange'
		,add: 'updatePositions'
		,remove: 'updatePositions'
		,childmove: 'updatePositions'
		,boxready: 'initialize'
	}
});
